platforms/extensibility/plugin-use/src/main/java/org/gradle/plugin/use/resolve/service/internal/ClientInjectedClasspathPluginResolver.java

```diff
@@ -16,12 +16,22 @@
 
 package org.gradle.plugin.use.resolve.service.internal;
 
+import org.gradle.api.internal.initialization.ScriptHandlerInternal;
 import org.gradle.plugin.use.resolve.internal.PluginResolver;
 
 import java.util.Collection;
 
 public interface ClientInjectedClasspathPluginResolver {
-    ClientInjectedClasspathPluginResolver EMPTY = dest -> { };
+    ClientInjectedClasspathPluginResolver EMPTY = new ClientInjectedClasspathPluginResolver() {
+        @Override
+        public void collectResolversInto(Collection<? super PluginResolver> dest) {
+        }
+        @Override
+        public void prepareClassPath(ScriptHandlerInternal scriptHandler) {
+        }
+    };
 
     void collectResolversInto(Collection<? super PluginResolver> dest);
+
+    void prepareClassPath(ScriptHandlerInternal scriptHandler);
```

Comment:
A couple of comments here. 

1. This is missing javadoc. This method has a very generic name. What does `prepareClassPath` do?  When I am supposed to call it? Do I _need_ to call it? 

2. This looks like it _needs_ to be called. This is always called before `collectResolversInto`. TBH it really seems like they should be the same method. In order to get the resolvers to collect, I should be _required_ to pass these parameters. There should be no way to call `collectResovlersInto`, and potentially get a resolver in an invalid state, without passing in this parameter. 

3. This takes in a whole `ScriptHandlerInternal` instance but in practice does not need access to nearly all the methods on it. Really all this method needs is a way to create a detached configuration and a way to create dependencies. We should be able to inject a `DependencyFactoryInternal` into the constructor of `DefaultInjectedClasspathPluginResolver`, meaning this only needs a detached configuration factory. I think that by limiting the parameters to what this method actually needs will greatly enhance understanding and improve maintainability. Most importantly, it shows that this injected classpath resolver really doesn't have _any_ dependency on the script handler. It just so happens that the script handler is the only way to get a detached configuration in this part of the codebase.

Reply:
Yeah, this implementation was heavily influenced by [DefaultScriptClassPathResolver](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java), that has similar "init" methods.

But I agree, we can do better have. Init methods are confusing. I can move to `collectResovlersInto`.

I will also take a look if we can avoid passing `ScriptHandlerInternal` as a parameter.